aten/src/ATen/native/SegmentReduce.cpp

```diff
@@ -103,21 +103,114 @@ void _segment_reduce_cpu_kernel1(
+              } else if (
+                  reduction == SegmentReductionType::MEAN ||
+                  reduction == SegmentReductionType::SUM) {
+                initial_value = 0;
+              } else if (reduction == SegmentReductionType::MIN) {
+                initial_value = std::numeric_limits<scalar_t>::infinity();
+              } else if (reduction == SegmentReductionType::PROD) {
+                initial_value = 1;
+              }
+
+              // ===== step2: apply reduction
+              for (const auto j : c10::irange(segment_length)) {
+                int64_t data_index = outer_idx * data.stride(axis) * data.size(axis)
+                                     + (lengths_cum_sum + j) * data.stride(axis) + inner_idx;
+                const auto val = values_data[data_index];
+                if (reduction == SegmentReductionType::MAX) {
+                  initial_value = at::_isnan(val)
+                      ? val
+                      : std::max<scalar_t>(initial_value, val);
+                } else if (
+                    reduction == SegmentReductionType::MEAN ||
+                    reduction == SegmentReductionType::SUM) {
+                  initial_value = initial_value + val;
+                } else if (reduction == SegmentReductionType::MIN) {
+                  initial_value = at::_isnan(val)
+                      ? val
+                      : std::min<scalar_t>(initial_value, val);
+                } else if (reduction == SegmentReductionType::PROD) {
+                  initial_value = initial_value * val;
+                }
+              }
+
+              // ===== step3: finalize reduction
+              TORCH_CHECK(segment_length >= 0);
+
+              if (segment_length == 0 && !initial.has_value() &&
+                  reduction == SegmentReductionType::MEAN) {
+                initial_value = static_cast<scalar_t>(NAN);
+              } else if (
+                  reduction == SegmentReductionType::MEAN &&
+                  segment_length > 0 && !at::_isnan(initial_value)) {
+                initial_value = initial_value / segment_length;
+              }
+              int64_t output_index = outer_idx * output.stride(axis) * output.size(axis)
+                                     + dim_idx * output.stride(axis) + inner_idx;
+              output_data[output_index] = initial_value;
+            }
+            lengths_cum_sum += segment_length;
+          }
+        }
+      });
+}
+
 Tensor _segment_reduce_cpu_kernel(
     SegmentReductionType reduction,
     const Tensor& data,
     const Tensor& lengths,
     int64_t axis,
     const c10::optional<Scalar>& initial) {
-  int64_t segment_count = lengths.numel();
+  // data and lengths should be contiguous from the call to .contiguous in segment_reduce_kernel
+  TORCH_CHECK(data.is_contiguous(), "Expected data to be contiguous.");
+  TORCH_CHECK(lengths.is_contiguous(), "Expected lengths to be contiguous.");
+  axis = lengths.dim() - 1;
```

Comment:
(note for myself) FIXME: add logic that expands index like here (https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/cpu/segment_coo_cpu.cpp#L18-L21) perhaps in another kernel like _segment_csr_cpu_kernel that calls into this